---
title: "RNA-seq Pipeline"
author: "Ross Thorne"
output: html_notebook
---

This is an example RNA-seq analysis, for instruction. It is a simple expriment,
with two conditions: control (SIMM) and treatment (SIMA6). Each condition has
3 replicates.


```{r setup}
library(pacman) # Easy package management
pacman::p_load(tidyverse, edgeR, GenomicFeatures)
```


## Feature Counts Import

Gather the data to build the DGEList object. The counts file was produced using 
the `featurecounts` command from the `subread` package.

```{r import, tidy = T, results="hold"}
# Import the feature counts, check dataframe size
features <- read_tsv("data/nascent_rnaseq/SIMA6/counts/counts_hg19.txt", comment = "#")
dim(features)

# Set sensible sample names
names(features)[7:12] <- c("SIMA6_1", "SIMA6_2", "SIMA6_3", "SIMM_1", "SIMM_2", "SIMM_3")
# Order the columns so that control appears first - otherwise upregulation
# will look like downregulation!
colnames(features)
features %>%
  dplyr::select(
    Geneid:Length, 
    starts_with("SIMM"), 
    starts_with("SIMA6")) ->
  features
colnames(features)

# Make a grouping factor
group <- factor(c(1,1,1,2,2,2))
# Make a counts matrix with gene names as row names
features %>%
  column_to_rownames("Geneid") %>%
  .[6:11] %>%
  as.matrix() ->
  counts

# Construct the DGEList
dgelist <- DGEList(counts = counts, group = group, genes = features[1:6])
```

## Filtering

Removing genes with less than 1 count per million appears to be the standard
for filtering out "lowly expressed genes". We also remove genes for which there
are no reads for one or more replicates.

```{r filtering}
cpm <- cpm(dgelist)
# Filter out genes that have less than 1 cpm per sample group
# Additionally, there should be 'expression' in at least 3 libraries
cpm.filter <- rowSums(cpm > 1) >= min(table(group))
dgelist <- dgelist[cpm.filter, ]

# Recalculate library size for the filtered counts
dgelist$samples$lib.size <- colSums(dgelist$counts)
# Check the dimensions of filtered list
dim(dgelist)
```

## Normalisation

```{r normalise}
# Normalise for RNA composition
dgelist.norm <- calcNormFactors(dgelist)
design <- model.matrix( ~ group)
```

## Estimate dispersion

```{r dispersion}
set.d <- estimateGLMCommonDisp(dgelist.norm, design, verbose = T)
set.d <- estimateGLMTrendedDisp(set.d, design)
set.d <- estimateGLMTagwiseDisp(set.d, design)

# Fit the model
fit <- glmFit(set.d, design)
testglm <- glmLRT(fit, coef=2)
len.fdr <- length(which(topTags(testglm, n=nrow(dgelist.norm))[[1]][,"FDR"]<0.05))
```

## Include CPMC

Reattach Counts per Million (CPM) values for each sample.

```{r}
counts_cpm <- dgelist.norm %>% cpm() %>%
  as_data_frame() %>%
  mutate(Geneid = rownames(dgelist.norm))
```

## Differentially Expressed Genes

Extract the top DE tags.

```{r export_results}
top.tags <- topTags(testglm, n = len.fdr)$table %>% 
  left_join(., y = counts_cpm, by = "Geneid")
all.tags <- topTags(testglm, n = nrow(dgelist.norm))$table %>% 
  left_join(., y = counts_cpm, by = "Geneid")

top.tags %>%
  dplyr::select(Geneid, SIMM_1:SIMA6_3, Length:FDR) %>%
  write_tsv(path = "output/nascent/SIMA6/SIMA6_hg19_ref_top.txt")

all.tags %>%
  dplyr::select(Geneid, SIMM_1:SIMA6_3, Length:FDR) %>%
  write_tsv(path = "output/nascent/SIMA6/SIMA6_hg19_ref_all.txt")

```
